launch_testing/launch_testing/__init__.py

```diff
@@ -53,14 +53,15 @@ def __init__(
         self.name = name
         self.launch_descriptor = launch_descriptor
         self.expected_lines = expected_lines
-        self.expected_output = b'\n'.join(self.expected_lines)
+        self.expected_output = b'\n'.join(self.expected_lines) + b'\n'
         self.left_over_stdout = b''
         self.left_over_stderr = b''
         self.stdout_data = io.BytesIO()
         self.stderr_data = io.BytesIO()
         self.regex_match = regex_match
         self.exit_on_match = exit_on_match
         self.matched = False
+        self.matched_exactly = False
```

Comment:
Just from reading the code I would expect `matched_exactly` to have a wrong value if `matched` becomes try early while processing the lines. As soon as that happens `on_stdout_lines` stops processing further lines and therefore never computes `matched_exactly` again.